src/sdk/pynni/nni/pbt_tuner/pbt_tuner.py

```diff
@@ -0,0 +1,337 @@
+            except nni.NoMoreTrialError:
+                #add credit
+                had_exception = True
+            if not had_exception:
+                result.append(res)
+        return result
+
+
+    def generate_parameters(self, parameter_id, **kwargs):
+        """
+        Generate parameters, if no trial configration for now, self.credit plus 1 to send the config later
+
+        Parameters
+        ----------
+        parameter_id : int
+            Unique identifier for requested hyper-parameters.
+            This will later be used in :meth:`receive_trial_result`.
+        **kwargs
+            Not used
+
+        Returns
+        -------
+        dict
+            One newly generated configuration
+
+        """
+        if self.pos == self.population_size-1:
+            logger.debug('Credit added by one in parameters request')
+            self.credit += 1
+            self.param_ids.append(parameter_id)
+            raise nni.NoMoreTrialError('no more parameters now.')
+        trial_info = self.population[self.pos]
+        trial_info.parameter_id = parameter_id
+        self.running[parameter_id] = trial_info
+        self.pos += 1
+        logger.info('generate parameter : %s', trial_info.hyper_parameters)
+        return trial_info.hyper_parameters
+
+
+    def receive_trial_result(self, parameter_id, parameters, value, **kwargs):
+        """
+        Receive trial's result. if the number of finished trials equals self.population_size, start the next epoch to
+        train the model.
+
+        Parameters
+        ----------
+        parameter_id : int
+            Unique identifier of used hyper-parameters, same with :meth:`generate_parameters`.
+        parameters : dict
+            Hyper-parameters generated by :meth:`generate_parameters`.
+        value : dict
+            Result from trial (the return value of :func:`nni.report_final_result`).
+        """
+        logger.info('get one trial result')
+        value = extract_scalar_reward(value)
+        if self.optimize_mode == OptimizeMode.Minimize:
+            value = -value
+        trial_info = self.running.pop(parameter_id, None)
+        trial_info.score = value
+        self.finished.append(trial_info)
+        self.finished_trials += 1
+        if self.finished_trials == self.population_size:
+            logger.info('next epoch')
+            self.epoch += 1
+            self.population = []
+            self.running = {}
+            #exploit and explore
+            self.finished = sorted(self.finished, key=lambda x: x['score'], reverse=True)
+            cutoff = int(np.ceil(self.fraction * len(self.finished)))
+            tops = self.finished[:cutoff]
+            bottoms = self.finished[self.finished_trials - cutoff:]
+            for bottom in bottoms:
+                top = np.random.choice(tops)
+                exploit_and_explore(bottom, top, self.factors, self.epoch)
+            for trial in self.finished:
```

Comment:
why not `for top in tops` here?

Reply:
actually, we should exploit each "bottom" trial so that here we loop through them